Bugfix for Issue #73 - Uninstall Command bug

eyallampel1 opened the pull request:
## Overview
This pull request introduces a fix for issue #73, where the `cocmd uninstall` command did not remove the associated package files from the `~/.cocmd/runtime/` directory. The fix ensures that the package directory is deleted from the filesystem after the package is uninstalled from cocmd's internal management system.

## Changes
- Modified the `remove_package` function to include a filesystem deletion call.
- Added error handling to manage potential deletion failures and provide feedback.

## How to Test
1. Run `cocmd install zsh`.
2. Run `cocmd uninstall zsh`.
3. Verify that `ls -ltr ~/.cocmd/runtime/zsh` indicates the directory no longer exists.

## Impact
This change ensures that uninstalling packages via cocmd is a clean operation, leaving no unnecessary files or directories behind, thus maintaining system hygiene.

Closes #73

eyallampel1 commented:
I've made the requested changes and pushed them. 
Please review .  
 it appears that all tests are passing successfully

mzsrtgzr2 commented on src/core/packages_manager.rs:
```diff
@@ -27,16 +27,23 @@ impl PackagesManager {
+                get_provider(&uri, &self.settings.runtime_dir).map_err(|e| e.to_string())?;
+            let package_dir = provider.get_installation_path();
+
+            if let Err(e) = std::fs::remove_dir_all(package_dir) {
```
can we make sure it will only remove in the runtime folder and not in other locations?
i'm asking because it's possible to add packages from any local path on a computer, for example:
`cocmd install /some/path/in/computer`
and if you have `cocmd.yaml` there it will add it. 
i think we need to make sure the provider is not of type [LocalPackageProvider](https://github.com/cocmd/cocmd/blob/master/src/package_provider/local.rs#L32)

mzsrtgzr2 replied:
thanks!

mzsrtgzr2 commented:
also @eyallampel1 i found out today there is a problem when you first delete manually from the runtime
```
➜  ~ rm -rf ~/.cocmd/runtime/python
➜  ~ cocmd uninstall python
[2023-11-06T07:32:08Z ERROR cocmd::core::package] Package Path /Users/mosher/.cocmd/runtime/python does not exist.
[2023-11-06T07:32:08Z INFO  cocmd::cmd::uninstall] Uninstalling package "python"
thread 'main' panicked at 'Unable to get name for /Users/mosher/.cocmd/runtime/python', src/core/package.rs:91:17
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

tell me if it's getting complicated we'll work together

eyallampel1 commented:
**Update on Pull Request for Issue #73**

I have addressed the concerns raised in the previous review. The updated implementation now ensures that only packages within the `~/.cocmd/runtime/` directory are removed by the uninstall command. This is achieved by checking the provider type and confirming it's not a `LocalPackageProvider` before proceeding with the deletion.

Here's a summary of the changes made:

- Incorporated the use of `LOCAL_PROVIDER` constant to ensure consistency and avoid hard-coded strings.
- Updated the `remove_package` function to perform a provider type check to safeguard against removing directories outside the runtime folder.
- Enhanced error handling to provide clear feedback in case of issues during the filesystem deletion process.

**Testing Steps:**
1. Run `cocmd install /some/path/in/computer`.
2. Run `cocmd uninstall zsh`.
3. Verify that `ls -ltr ~/.cocmd/runtime/zsh` confirms the directory has been removed, and no other locations are affected.

These changes reinforce the safety and reliability of the uninstall process, ensuring that only the intended directories are affected, thereby maintaining the integrity of the user's filesystem.

I have committed the updated code, and all tests are passing successfully. Please review the changes
